[docs] Update docs on code-review process

banach-space opened the pull request:
Clarify what to do in cases where multiple reviewers commented on a
patch, but only one reviewer explicitly accepted.

This is mostly to standardize what the expectations are as folks tend to
approach this differently.

banach-space committed:
[docs] Update docs on code-review process

Clarify what to do in cases where multiple reviewers commented on a
patch, but only one reviewer explicitly accepted.

This is mostly to standardize what the expectations are as folks tend to
approach this differently.

banach-space commented:
Follow-on from a discussion here: https://github.com/llvm/llvm-project/pull/111533

My assumption is that this is documenting what many of us already practice, so hopefully this is nothing particularly new. Please let me know if you feel otherwise.

Any suggestion who else to include in the review?

kuhar commented on llvm/docs/Contributing.rst:
```diff
@@ -105,9 +105,16 @@ get the attention of potential reviewers by CC'ing them in a comment -- just
-have been addressed and a reviewer accepts the patch with a `Looks good to me` or `LGTM`.
-Once that is done the change can be committed. If you do not have commit
+updated version of your patch. This cycle continues until all requests and
+comments have been addressed and the reviewer accepts the patch with a `Looks
+good to me` or `LGTM`. With multiple active reviewers (i.e. reviewers who left
```
Should we s/LGTM/approval/ (as in the button in the github ui)? I think 'LGTM' might be the language coupled to some older iteration of our code review process that feels a bit out of place now

nikic reviewed (changes requested):
This document is just an overview, we have detailed documentation of the code review policy in https://llvm.org/docs/CodeReview.html, which is also already linked. 

In particular, https://llvm.org/docs/CodeReview.html#lgtm-how-a-patch-is-accepted explains that it is the responsibility of the approving reviewer to consider other feedback on the patch and indicate if their approval is conditional.

joker-eph commented on llvm/docs/Contributing.rst:
```diff
@@ -105,9 +105,16 @@ get the attention of potential reviewers by CC'ing them in a comment -- just
+  leave a short justification.
+Once a change has been approved, it can be committed. If you do not have commit
 access, please let people know during the review and someone should commit it
 on your behalf.
```
Instead of adding more text here, I think we should trim it down and link to the dedicated page: https://llvm.org/docs/CodeReview.html

Right now you're evolving the text in a direction that contradicts what this other page is saying I believe.

banach-space replied:
> I think we should trim it down and link to the dedicated page

Agreed, I think that repeating similar information in multiple places might be confusing. I went ahead and deleted a paragraph from Contributing.rst to avoid that.

banach-space committed:
fixup! [docs] Update docs on code-review process

Revert changes in llvm/docs/Contributing.rst, update llvm/docs/CodeReview.rst instead.

banach-space committed:
fixup! fixup! [docs] Update docs on code-review process

Remove paragraph from Contributing.rst

joker-eph commented on llvm/docs/CodeReview.rst:
```diff
@@ -150,14 +150,18 @@ almost always associated with a message containing the text "LGTM" (which
 
+If new comments are posted after the patch has been approved (but not yet
+merged), these need to be addressed and all new changes have to be reviewed and
+approved by a reviewer.
```
The second part of the sentence is a bit strongly worded, what do you mean to cover here?

kuhar replied:
+1, I don't think we want to effectively require a re-approval after every push.
For example, when a new reviewer comes and add a typo/nit comment, I'd expect that the author can fix that and merge based on the previous approval given how insignificant the comment (and subsequent fix) is. I think this is a standard practice today.

banach-space replied:
I've expanded this paragraph. @Jakub, hopefully that meets your criteria (which I mostly agree with).

> what do you mean to cover here?

What should happen when there's a "LGTM", but then another reviewer leaves new comments afterwards? I wanted to clarify that the original "LGTM" wouldn't apply to the new comments (unless, like Jakub points out, those would be  basic nits etc).

> For example, when a new reviewer comes and add a typo/nit comment, I'd expect that the author can fix that and merge based on the previous approval given how insignificant the comment (and subsequent fix) is. 

This is assuming that there won't be more substantial comments following the nits - is it a valid assumption? I guess the reviewer should make that clear, e.g. "Just some nits, no need to wait for my approval".

banach-space replied:
>  but I don't see how "all new changes have to be reviewed " is addressing what you describe here ("another reviewer leaves new comments afterwards").

That's been updated since :)

joker-eph reviewed (approved):
LG overall, I have one comment though, thanks!

banach-space committed:
fixup! fixup! fixup! [docs] Update docs on code-review process

Expand the new paragraph

kuhar commented on llvm/docs/CodeReview.rst:
```diff
@@ -150,14 +150,21 @@ almost always associated with a message containing the text "LGTM" (which
+merged), these need to be addressed following similar process as outlined
+above. Specifically, a reviewer should confirm that all feedback has been
+addressed before a patch is merged, including the newly posted comments.
+Exceptions apply - e.g. there's no need to confirm that a comment requesting a
```
To follow the style above

```suggestion
Exceptions apply, e.g., there's no need to confirm that a comment requesting a
```

kuhar commented on llvm/docs/CodeReview.rst:
```diff
@@ -150,14 +150,21 @@ almost always associated with a message containing the text "LGTM" (which
+above. Specifically, a reviewer should confirm that all feedback has been
+addressed before a patch is merged, including the newly posted comments.
+Exceptions apply - e.g. there's no need to confirm that a comment requesting a
+typo to be fixed has been addressed (this should be evident from the code).
```
I'd make it less specific to allow for some wiggle room we've traditionally had

```suggestion
minor issue (like a typo or other 'nit') to be fixed has been addressed (this should be evident from the code).
```

preames commented on llvm/docs/CodeReview.rst:
```diff
@@ -150,14 +150,21 @@ almost always associated with a message containing the text "LGTM" (which
 you are fairly certain that a particular community member will wish to review,
 even if that person hasn't done so yet.
 
+If new comments are posted after the patch has been approved (but not yet
```
This advise talks about action the reviewer should take, that seems backwards here.  Unless I'm misreading, the problematic situation here is a prior LGTM followed by changes by a *different* reviewer and the author landing the patch without addressing those changes.

I'd suggest something along the lines of the following:
Only approval from a single reviewer is required; however the patch author is always responsible for making sure that all feedback on a review has been properly addressed or responded to.  If additional feedback is provided after acceptance (by the same reviewer or another), the author should use their best judgement in deciding whether that feedback can be incorporated into the change without comment (say a typo) or requires further review discussion.  Similarly, the patch author is responsible for incorporating post commit feedback on the change.  

I'd put this paragraph just above the one talking about unqualified LGTM

banach-space replied:
>  Unless I'm misreading, the problematic situation here is a prior LGTM followed by comments by a different reviewer and the author landing the patch without addressing those comments.

Yes

>  (edit: for typo fix)

No, for more substantial changes. For "typos", I added an exception.

>  If additional feedback is provided after acceptance (by the same reviewer or another), the author should use their best judgement in deciding whether that feedback can be incorporated into the change without comment (say a typo) or requires further review discussion.

To me this sounds that any comments "post LGTM" are "second category" (i.e. "for the author to decide whether to address them or not" vs "need to be addressed"). I'm suggesting that we should treat such comments as any other comments prior to LGTM (i.e., "need to be addressed"). I feel that that otherwise we are discouraging people from reading/commenting on patches post LGTM.

preames replied:
> > Unless I'm misreading, the problematic situation here is a prior LGTM followed by comments by a different reviewer and the author landing the patch without addressing those comments.
> 
> Yes
> 
> > (edit: for typo fix)
> 
> No, for more substantial changes. For "typos", I added an exception.

(You're responding to a note I made about editting my text to fix a typo...)
 
> > If additional feedback is provided after acceptance (by the same reviewer or another), the author should use their best judgement in deciding whether that feedback can be incorporated into the change without comment (say a typo) or requires further review discussion.
> 
> To me this sounds that any comments "post LGTM" are "second category" (i.e. "for the author to decide whether to address them or not" vs "need to be addressed"). I'm suggesting that we should treat such comments as any other comments prior to LGTM (i.e., "need to be addressed"). I feel that that otherwise we are discouraging people from reading/commenting on patches post LGTM.

If you have suggested wording, feel free to suggest.  I don't read this the way you do, particular following the prior sentence that you trimmed.

banach-space replied:
Hi @preames , thanks for that suggestion and sorry for the delay - I had to step away and re-think this a bit. I do want to incorporate your feedback, but also want to make sure we avoid duplication and keep this document succinct. So I trimmed your suggestion a bit as per the comments below (hope this won't come across as nit-picking).

**Point 1:**
> [From Phillip's suggestion]
> Only approval from a single reviewer is required; 

This point is already mentioned earlier in the document: https://github.com/llvm/llvm-project/blob/c2063de1593610eda0f4de33c3b89324642ed54c/llvm/docs/CodeReview.rst?plain=1#L148-L150
 
**Point 2:**
> [From Phillip's suggestion]
> however the patch author is always responsible for making sure that all feedback on a review has been properly addressed or responded to. 

Isn't this a contradicting one existing policy?
https://github.com/llvm/llvm-project/blob/c2063de1593610eda0f4de33c3b89324642ed54c/llvm/docs/CodeReview.rst?plain=1#L152-L159

Specifically:

> it is the responsibility of the reviewer to have reviewed all of the discussion and feedback from all reviewers ensuring that all feedback has been addressed and that all other reviewers will almost surely be satisfied with the patch being approved.

I'd just leave this out.

**Point 3:**

> [From Phillip's suggestion]
> Similarly, the patch author is responsible for incorporating post commit feedback on the change.

This is already covered here: https://github.com/llvm/llvm-project/blob/c2063de1593610eda0f4de33c3b89324642ed54c/llvm/docs/CodeReview.rst?plain=1#L46-L49

**Point 4:**

Which leaves this bit from Phillip's original suggestion (emphasis mine):
> [From Phillip's suggestion]
If additional feedback is provided after acceptance (by the same reviewer or another), **the author should use their best judgement** in deciding whether that feedback can be incorporated into the change without comment (say a typo) or requires further review discussion. 

This a bit softer requirement compared to what I added "these need to be addressed" and sounds more along the lines of what @kuhar would prefer? Let me use that, but let me add small clarification at the end. Let me know what you think.

preames replied:
You have seemingly managed to complete miss the spirit of my original suggestion, and are misreading the current policy in my view.  If the document implies that the author is not responsible for incorporating feedback (because that's only the reviewers responsibility), the document is simply wrong.  I don't read it that way, but if you do, stop and fix that wording before working on this smaller item.

banach-space replied:
> You have seemingly managed to complete miss the spirit of my original suggestion

That's not intentional. Would you be able to share a diff so that I better understand what you had in mind? 

> If the document implies that the author is not responsible for incorporating feedback

I don’t believe the document implies that. I think it’s fairly clear that:

1. The author is responsible for addressing/incorporating the feedback, which is already covered under "Acknowledge All Reviewer Feedback" (so perhaps we can leave that as is).
2. The reviewer, when posting an unqualified LGTM, is responsible for ensuring that all feedback has been addressed (i.e., “I've reviewed all comments and confirm they’ve been addressed").

The existing wording covers those points well. The part that is unclear IMO is what happens **after LGTM** and **before commit** when new comments are posted. Hopefully, the additional paragraph I proposed clarifies that.

banach-space committed:
fixup! fixup! fixup! fixup! [docs] Update docs on code-review process

Incorporate PR feedback

nikic commented on llvm/docs/Contributing.rst:
```diff
@@ -102,21 +102,9 @@ will not be able to select reviewers in such a way, in which case you can still
-For more information on LLVM's code-review process, please see :doc:`CodeReview`.
+If you do not have commit access, please let people know during the review and
+someone should commit it on your behalf once it has been accepted. For more
+information on LLVM's code-review process, please see :doc:`CodeReview`.
```
I feel like this has cut out too much. E.g. I think the "ping" convention should still be mentioned here. This is quite relevant to new contributors, who otherwise don't need to read the full scope of the code review documentation.

banach-space replied:
I can leave it if you feel strongly about it, but IMO this is duplicating information and that can be confusing. Sometimes less is more and also - why not _encourage_ people to read our policy instead? it's an important document and it's there for a reason.

banach-space replied:
I'm keen to progress this, so I've just restored that note. Thanks!

banach-space committed:
fixup! fixup! fixup! fixup! fixup! [docs] Update docs on code-review process

Restore a note on weekly pings.

banach-space commented:
Gentle ping :)

joker-eph commented:
Still LGTM, but we should hear back from @nikic / @preames.

kuhar commented on llvm/docs/CodeReview.rst:
```diff
@@ -150,14 +150,20 @@ almost always associated with a message containing the text "LGTM" (which
+If additional feedback is provided after acceptance (by the same reviewer or
+another), the author should use their best judgement in deciding whether that
+feedback can be incorporated into the change without comment (say a typo) or
+requires further review discussion. More substantial comments (e.g. about the
```
Let's keep this consistent with the style used above
```suggestion
requires further review discussion. More substantial comments (e.g., about the
```

kuhar reviewed (approved):
LGTM % nit

banach-space committed:
fixup! fixup! fixup! fixup! fixup! fixup! [docs] Update docs on code-review process

Add missing coma

banach-space commented:
Ping @nikic and @preames - are you OK with me landing this in its current form? Thanks!

nikic reviewed (approved):
LGTM

banach-space commented:
Hey folks, thank you for all the reviews! If there are no new comments, I will land this tomorrow.
